### DISCUSSION TITLE
CSR과 SSR의 차이는 무엇인가요?

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#23] [프론트 기술 질문 공간] [answered]
Client-side-rendering vs Server-side-rendering 

SSG (Static-Site-Generation) 도 가능하면 비교해주세요

**Comment by WooGie911** [comment]
CSR(Client-Side-Rendering) 과 SSR(Server-Side-Rendering) 의 가장 큰 차이점은 
렌더링이 어디서 일어나느냐. 즉 클라이언트에게 전달되는 방식이 다르다는 것입니다.

말그대로 CSR은 렌더링이 클라이언트, SSR은 서버에서 준비됩니다. 

유저가 웹페이지에 요청을 보냈을 때 클라이언트는 서버로 부터 HTML과 JS를 다운로드 받습니다. 
이때 SSR과 달리 유저는 아무것도 볼 수 없습니다. 

이후 데이터를 위한 API가 호출되고 유저들은 placeholder를 보게 됩니다.

이후 서버로 부터 API를 통해 받아온 데이터를 placeholder 자리에 넣어주고 웹페이지는 상호작용이 가능해집니다.

반면에 SSR은 유저가 웹페이지에 요청을 보냈을 때 서버로 부터 즉시 렌더링이 가능한 HTML을 받고 즉시 렌더링 됩니다. 
허나 JS가 읽히기 전이기에 조작은 불가능합니다.

이후 클라이언트가 JS를 다운받는 동안 사용자 조작을 기억하고 있다가 
JS까지 컴파일 되면 사용자 조작이 실행되고 웹페이지는 상호작용이 가능해집니다.


CSR과 SSR의 가장 큰 차이점이라고 한다면 
JS가 다운로드 되는 동안 렌더링. 즉 사용자가 무언가를 보고 있을 수 있냐 없냐 입니다.

또다른 차이점이라면 CSR은 첫 페이지 로딩시간이 오래걸립니다. 
CSR의 경우 HTML, CSS와 모든 스크립트들을 한 번에 불러오는 반면 
SSR은 필요한 부분의 HTML과 스크립트만 불러오기 때문에 평균적으로 SSR이 더 빠릅니다.

하지만 나머지 로딩 시간은 다릅니다. 
첫 페이지를 로딩한 후, 사이트의 다른 곳으로 이동하는 식의 동작을 한다면 
CRS는 이미 나머지 부분도 받아왔기에 빠르지만 
SSR은 첫페이지를 로딩한 과정을 다시 실행하기에 더 느립니다.

**Comment by michoball** [comment]
CSR은 홈페이지 접속시 화면 구성을 위해 간단한 html, css 파일만 받아옵니다.  그리고 나서 화면에 동적인 동작을 위한 js파일을 받아와서 화면에 보여주는 방식입니다.  그렇기 때문에  js 파일을 받아와서 실행하기 전에는 사용자는 아무런 화면을 볼 수 없습니다.

SSR은 CSR 과는 다르게 화면을 보여주기위한 data와 html 파일을 서버에서 만들어서 보내줍니다. 그리고 나서 상호작용을 위한 js 파일을 받아옵니다.  때문에 유저는 상호작용이 되지 않는 화면을 먼저 보게 되고 이후 상호작용이 가능하게 됩니다. 

CSR은 사용자가 처음 화면을 보는데까지 시간이 오래걸리고 SEO가 좋지 않다는 단점이 있지만 서버에 부담이 덜합니다. 
SSR은 첫화면 로딩이 빠르고 SEO가 좋지만 서버에 부담이 많이가고 페이지안에서 이동할 때 페이지가 리랜더링 되면서 화면이 깜박이는 현상이 있다는 단점이 있습니다. 

SSG는 미리 랜더링을 다 마친 정적파일을 배포한 방식입니다. 화면을 보여주기 위한 data 가 json과 같은 정적 파일에 들어있고 이 data와 html css js 파일을 가지고 build을 마친 파일을 서버에 저장해서 홈페이지 접속시 보여주는 방식입니다.  ( ? ) -> 보통 페이지내에서 api요청이 필요가 없는 블로그와 같은 페이지를 만들때 효율적입니다.

**Comment by skyhanull** [comment]
csr은 클라이언트 사이드 랜더링의 약자로 클라이언트 단에서 랜더링을 하는 것이고 ssr은 서버사이드 랜더링의 약자로 서버단에서 랜더링을 하는 것을 말합니다 
이 둘의 차이점으로 초기로딩속도는 csr이 ssr보다 느린데 이유는 csr은 html,css,js를 초기로딩 때 다운을 받기에 느리지만 ssr은 필요한 부분만 데이터를 받아 랜더링하기에 csr보다 빠릅니다
나머지 로딩 시간은 반대로 csr은 이미 초반에 다 불러왓기 때문에 시간이 더 짧고 ssr은 요청할 때마다 필요한 데이터를 가져와야 하기에 느립니다.

**Comment by BaekYuri** [comment]
csr은 클라이언트 단에서 렌더링이 되는 것이고, ssr은 서버 단에서 렌더링 준비를 하고 클라이언트에 전달하는 방식입니다.
초기 로딩 속도는 ssr이 csr보다 빠릅니다. ssr은 필요한 부분의 HTML과 스크립트를 준비하는 반면에 csr은 html, css, 모든 스크립트를 불러오기 때문입니다.
나머지 로딩 시간은 csr이 ssr보다 빠릅니다. csr은 초기 로딩 시 모두 불러오기 때문에 빠르고, ssr은 첫 페이지를 로딩하는 만큼 다시 걸립니다.

**Comment by bona373737** [comment]
브라우저 렌더링 방식의 차이.
CSR : 서버로부터 렌더링에 필요한 html,css,js파일을 전송받아 브라우저에서 렌더링 하는 방식.
react의 경우 JSX 문법을 사용해서 React 요소를 만들고 DOM에 렌더링 시켜서 그리기때문에 서버로 부터 전달받은 html파일은 비어있다. 때문에 js파일이 다 실행되기 전까지는 빈화면이 보입니다. 

SSR : 서버가 화면을 그려주는 주체로서 화면에 보여질 내용을 서버쪽에서 그려서 완성된 html문서를 클라이언트에 전송한다.

**Comment by michoball** [comment, marked as answer]
## CSR 
### 클라이언트에서 웹 페이지를 위한 모든 작업을 수행한다. 

1. 서버로부터 index.html을 받아온다 이때 여기에는 React 프로젝트에서 볼 수 있듯 app에 
필요한 js 링크 ( <div id='root' /> )만 담겨 있다. 
2.  이후 필요한 js 파일을 서버로부터 받아온다. 이때 app의 구동을 위한 모든 js 파일을 받아온다.
3. 추가로 필요한 data 가 있다면 서버에 요청해 받아온다. (data.json)
4. 받아온 모든 파일을 실행해 app을 구현한다.
5. 바로 화면이 보여지고 상호작용을 할 수 있는 웹 페이지를 보여준다.

단점: 첫 화면을 보기까지 시간이 많이 소요된다, SEO가 좋지 않다.

---

## SSR
### 서버 측에서 우선 화면을 띄우기 위한 html 파일을 만들어서 보내준다.

1. 서버에서 화면 구성에 필요한 data를 가지고 화면의 모습이 만들어진 html 파일을 보낸다. 
2. 이후 화면의 여러 element들과 상호작용을 할 때 필요한 js 파일을 받아온다. 
3. 받아온 js로 상호작용이 가능한 웹페이지가 보여진다.

단점:  첫 화면을 빨리 볼 수 있지만 상호작용을 하려면 약간 시간이 걸린다.  화면이동시 새로 화면 구성을 위한 data를 서버에서 받아와야하기 때문에 깜박거리는 현상이 발생한다, 사용자가 많은 페이지면 서버에 부담이 많이간다. 

---

## SSG 

 정적으로 모든 페이지에 대한 html 파일을 미리 서버에 빌드후 배포하여 사용자의 요청에 따라 이미 만들어진 페이지를 보낸다. ( 모든 것을 정적으로 가지는 것은 아니고 필요에 따라 동적인 작업을 위한 js 파일을 가질 수 있다. )

---

## 결론

### 사용자가 보는 화면을 렌더링하는 위치에 따라서 CSR과 SSR이 나눠집니다. CSR은 사용자 측에서 서버로부터 받아온 html, css, js 파일을 가지고 웹페이지 화면을 만들고 SSR은 서버에서  화면 구성에 필요한 파일로 미리 화면을 만들어서 사용자에게 보낸다는 차이가 있습니다. 
### CSR은 사용자에서 모든 것을 해결하기 때문에 웹페이지의 첫 화면을 보는데까지 시간이오래걸리고 아무 정보가 없는 html 파일을 받아오기 때문에 SEO가 좋지 않습니다. 그대신 화면이 보이면 모든 상호작용을 할 수 있습니다. 
### 반면에 SSR은 서버에서 미리 화면을 만들어서 보내주기 때문에 첫 화면을 빨리 볼 수 있고 html에 많은 정보가 들어있어 SEO가 좋지만,  이후에 상호작용을 위한 js 파일을 나중에 받아오기 때문에 상호작용을 하는데까지 시간이 걸립니다. 

### SSG는 정적인 페이지를 구현할 때 많이 사용합니다.  이미 빌드된 파일을 받아오기 때문에 페이지 로딩도 빠르고 SEO도 좋습니다. 다만 상호작용을 많이 해야하는 웹 페이지에는 적합하지 않을 수 있습니다.
